### ISSUE TITLE
Compilation options of a multi-output model: multiple losses & loss weighting

### ISSUE THREAD

**Issue by GKalliatakis** [keras-team/keras#10306]
As described in the Keras handbook -_Deep Learning with Pyhton_-, for a multi-output model we need to specify different loss functions for different heads of the network. But because gradient descent requires you to minimize a scalar, you must combine these losses into a single value in order to train the model. 

Very imbalanced loss contributions will cause the model representations to be optimized preferentially for the task with the largest individual loss, at the expense of the other tasks. To remedy this, you can assign different levels of importance to the loss values in their contribution to the final loss. This is useful in particular if the losses’ values use different scales.

Can anyone help with the following:

I've got a five-output model as described in #10120. 
The outputs of the model are the following:
1. emotion (multiclass, multilabel classification)
2. valence (regression)
3. arousal (regression)
4. dominance (regression)
5. age (multiclass classification)

I am using the following : 

```python
    losses_list = {'EMOTIONS': 'binary_crossentropy',
                   'VALENCE': 'mse',
                   'AROUSAL': 'mse',
                   'DOMINANCE': 'mse',
                   'AGE': 'categorical_crossentropy'}
```

```python
    losses_weights = {'EMOTIONS': 1.0,
                      'VALENCE': 0.025,
                      'AROUSAL': 0.025,
                      'DOMINANCE': 0.025,
                      'AGE': 0.45}
```

```python
    metrics ={'EMOTIONS': 'crossentropy',
              'VALENCE': 'mse',
              'AROUSAL': 'mse',
              'DOMINANCE': 'mse',
              'AGE': 'categorical_accuracy'}
```

Can anyone comment on this ? 
Are those the right loss functions? Are those the right weights and are those metrics properly set?

**Comment by brge17** [comment]
7900/60000 [==>...........................] - ETA: 44s - loss: 1.6085 - classification_loss: 1.3906 - autoencoder_loss: 0.4358 - classification_acc: 0.5711 - autoencoder_mean_squared_error: 0.1194

**Comment by GKalliatakis** [comment]
Many thanks for that @brge17. 

Can you please elaborate on 
```python 
loss_weights={'classification': 1.0, 'autoencoder': 0.5}
```

and how you've chosen those weights (floats) ?

Also why  do you use 2 metrics for the autoencoder?
```python 
metrics={'classification': 'accuracy', 'autoencoder': ['binary_crossentropy', 'mse']}
```

**Comment by brge17** [comment]
It's a toy problem, the loss weights are meaningless but show you how to use them.

Same with the metrics, just showing you how to use them.

**Comment by tie304** [comment]
I'm also having the same struggle . @GKalliatakis  Have you had any progress with this?

**Comment by GKalliatakis** [comment]
> Very imbalanced loss contributions will cause the model representations to be optimized preferentially for the task with the largest individual loss, at the expense of the other tasks. To remedy this, you can assign different levels of importance to the loss values in their contribution to the final loss. This is useful in particular if the losses' values use different scales. For instance, the mean squared error (MSE) loss used for the age-regression task typically takes a value around 3–5, whereas the cross-entropy loss used for the gender-classification task can be as low as 0.1. In such a situation, to balance the contribution of the different losses, you can assign a weight of 10 to the crossentropy loss and a weight of 0.25 to the MSE loss.

The above quote is taken from the Deep Learning with Python book. This is actually the only bit that I've found online that gives an actual example. I assume you need to figure out the values that each loss typically take and then assign weights accordingly.

**Comment by soans1994** [comment]
hello,

i want to train a model to predict heatmap (mean square error loss) and binary segmentation (binary cross entropy loss).
i tried to train model using multi branch (2 branch duplicates for 2 output). but the the final output will favour for only one type of output. 
For example when i train using model.fit with equal loss weights, the output is good for heatmap, but binary mask output is wrong and gives pixels 1 for the regions similar to heatmaps.
And when i train using gradtape loop, the output is good for segmentation mask, but heatmaps are wrong and looks like masks.

how can i solve this, please give me your suggestion.

thank you

**Comment by g-jindal2001** [comment]
@soans1994 Any updates on this issue? I am too facing the same issue

**Comment by soans1994** [comment]
@g-jindal2001 
apparently we cannot train model for 2 entirely different tasks like heatmaps and binary mask since the weights cannot learn 2 different tasks. 
In my case , i used 2 outputs at the end of the model by using 2 final dense layers like below. this is wrong since all weights are shared for these 2 outputs. and cannot learn both. Note at the end both output and output2 takes x (same branch data)
```
from tensorflow.keras.layers import Input, concatenate, Dense, Flatten, BatchNormalization, Dropout, ReLU, Conv2D, Reshape, MaxPooling2D, Conv2DTranspose
from tensorflow.keras.models import Model, Sequential
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input
import tensorflow as tf
from tensorflow.keras.regularizers import l2

input_shape = (256,256,3)
Nkeypoints = 15

def model(input_shape):
    H = input_shape[0]
    W = input_shape[1]
    def downsample_block(x, block_num, n_filters, pooling_on=True):

        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv1")(x)
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv2")(x)
        skip = x

        if pooling_on is True:
            x = MaxPooling2D(pool_size=(2, 2), strides=2, padding='valid', name="Block" + str(block_num) + "_Pool1")(x)

        return x, skip

    def upsample_block(x, skip, block_num, n_filters):

        x = Conv2DTranspose(n_filters, kernel_size=(2, 2), strides=2, padding='valid', activation='relu',
                            name="Block" + str(block_num) + "_ConvT1")(x)
        x = concatenate([x, skip], axis=-1, name="Block" + str(block_num) + "_Concat1")
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv1")(x)
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv2")(x)

        return x

    input = Input(input_shape, name="Input")

    # downsampling
    x, skip1 = downsample_block(input, 1, 64)
    x, skip2 = downsample_block(x, 2, 128)
    x, skip3 = downsample_block(x, 3, 256)
    x, skip4 = downsample_block(x, 4, 512)
    x, _ = downsample_block(x, 5, 1024, pooling_on=False)

    # upsampling
    x = upsample_block(x, skip4, 6, 512)
    x = upsample_block(x, skip3, 7, 256)
    x = upsample_block(x, skip2, 8, 128)
    x = upsample_block(x, skip1, 9, 64)

    output = Conv2D(17, kernel_size=(1, 1), strides=1, padding='valid', activation='linear', name="output")(x)#heatmap
    output2 = Conv2D(1, kernel_size=(1, 1), strides=1, padding='valid', activation='sigmoid', name="output2")(x)  # binary mask
    model = Model(inputs=input, outputs=[output, output2], name="Output")
    model.summary()
    return model
if __name__=="__main__":
    model(input_shape)

```
So, we have to build 2 branches same as first one. 2 branches weights to learn 2 different task and finally concatenate outputs.
note at the end output and output2 takes x and xx resepectivly.
```
from tensorflow.keras.layers import Input, concatenate, Dense, Flatten, BatchNormalization, Dropout, ReLU, Conv2D, Reshape, MaxPooling2D, Conv2DTranspose
from tensorflow.keras.models import Model, Sequential
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input
import tensorflow as tf
from tensorflow.keras.regularizers import l2

input_shape = (256,256,3)
Nkeypoints = 15

def model(input_shape):
    H = input_shape[0]
    W = input_shape[1]
    def downsample_block(x, block_num, n_filters, pooling_on=True):

        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv1")(x)
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv2")(x)
        skip = x

        if pooling_on is True:
            x = MaxPooling2D(pool_size=(2, 2), strides=2, padding='valid', name="Block" + str(block_num) + "_Pool1")(x)

        return x, skip

    def upsample_block(x, skip, block_num, n_filters):

        x = Conv2DTranspose(n_filters, kernel_size=(2, 2), strides=2, padding='valid', activation='relu',
                            name="Block" + str(block_num) + "_ConvT1")(x)
        x = concatenate([x, skip], axis=-1, name="Block" + str(block_num) + "_Concat1")
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv1")(x)
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block" + str(block_num) + "_Conv2")(x)

        return x

    def downsample_block2(x, block_num, n_filters, pooling_on=True):

        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block2" + str(block_num) + "_Conv1")(x)
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block2" + str(block_num) + "_Conv2")(x)
        skip = x

        if pooling_on is True:
            x = MaxPooling2D(pool_size=(2, 2), strides=2, padding='valid', name="Block2" + str(block_num) + "_Pool1")(x)

        return x, skip

    def upsample_block2(x, skip, block_num, n_filters):

        x = Conv2DTranspose(n_filters, kernel_size=(2, 2), strides=2, padding='valid', activation='relu',
                            name="Block2" + str(block_num) + "_ConvT1")(x)
        x = concatenate([x, skip], axis=-1, name="Block2" + str(block_num) + "_Concat1")
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block2" + str(block_num) + "_Conv1")(x)
        x = Conv2D(n_filters, kernel_size=(3, 3), strides=1, padding='same', activation='relu',
                   name="Block2" + str(block_num) + "_Conv2")(x)

        return x

    input = Input(input_shape, name="Input")

    # downsampling
    x, skip1 = downsample_block(input, 1, 64)
    x, skip2 = downsample_block(x, 2, 128)
    x, skip3 = downsample_block(x, 3, 256)
    x, skip4 = downsample_block(x, 4, 512)
    x, _ = downsample_block(x, 5, 1024, pooling_on=False)

    # upsampling
    x = upsample_block(x, skip4, 6, 512)
    x = upsample_block(x, skip3, 7, 256)
    x = upsample_block(x, skip2, 8, 128)
    x = upsample_block(x, skip1, 9, 64)
######################################################
    # downsampling seg
    xx, sskip1 = downsample_block2(input, 1, 64)
    xx, sskip2 = downsample_block2(xx, 2, 128)
    xx, sskip3 = downsample_block2(xx, 3, 256)
    xx, sskip4 = downsample_block2(xx, 4, 512)
    xx, _ = downsample_block2(xx, 5, 1024, pooling_on=False)

    # upsampling seg
    xx = upsample_block2(xx, sskip4, 6, 512)
    xx = upsample_block2(xx, sskip3, 7, 256)
    xx = upsample_block2(xx, sskip2, 8, 128)
    xx = upsample_block2(xx, sskip1, 9, 64)

    output = Conv2D(17, kernel_size=(1, 1), strides=1, padding='valid', activation='linear', name="output")(x)#heatmap
    output2 = Conv2D(1, kernel_size=(1, 1), strides=1, padding='valid', activation='sigmoid', name="output2")(xx)  # mask
    model = Model(inputs=input, outputs=[output, output2], name="Output")
    model.summary()
    return model
if __name__=="__main__":
    model(input_shape)

```
